src/Compilers/Extension/CompilerPackage.cs

```diff
@@ -4,25 +4,30 @@
 using System.Collections.Generic;
 using System.IO;
 using System.Reflection;
+using System.Threading;
 using EnvDTE;
 using Microsoft.VisualStudio.Shell;
 using Microsoft.VisualStudio.Shell.Interop;
+using Task = System.Threading.Tasks.Task;
 
 namespace Roslyn.Compilers.Extension
 {
-    [ProvideAutoLoad(UIContextGuids.SolutionExists)]
-    public sealed class CompilerPackage : Package
+    [ProvideAutoLoad(UIContextGuids.SolutionExists, PackageAutoLoadFlags.BackgroundLoad)]
+    [PackageRegistration(UseManagedResourcesOnly = true, AllowsBackgroundLoading = true)]
+    public sealed class CompilerPackage : AsyncPackage
     {
         public static string RoslynHive = null;
         
-        protected override void Initialize()
+        protected override async Task InitializeAsync(CancellationToken cancellationToken, IProgress<ServiceProgressData> progress)
         {
-            base.Initialize();
+            await base.InitializeAsync(cancellationToken, progress).ConfigureAwait(true);
```

Comment:
so InitializaAsync is no op. and it switch to main thread right away, and do all works on UI thread. so only thing it buy is GetServiceAsync? but it is querying SLocalRegistry which seems free-thread native service (https://devdiv.visualstudio.com/DevDiv/Default/_search?text=ILocalRegistry2&type=code&lp=code-Team&filters=ProjectFilters%7BDevDiv%7DRepositoryFilters%7BVS%7DBranchFilters%7Blab%2Fvsumain%7DCodeElementFilters%7Bdecl%7D&result=DefaultCollection%2FDevDiv%2FVS%2FGBlab%2Fvsumain%2F%2Fsrc%2Fappid%2Flib%2Fclrhosting.cpp&_a=contents) so feels like it will be pretty quick.

so we are converting this just to be consistent?

Reply:
> so we are converting this just to be consistent?

Yes, the pattern was _intended_ to minimize the behavior change from what we currently do, though it appears some subtleties came in and broke things anyway...

Reply:
What broke?
And certainly, if you're confident that what you're doing is free-threaded, you can drop the SwitchToMainThreadAsync call (or move the free-threaded code above that call).

---
In reply to: [188655773](https://github.com/dotnet/roslyn/pull/26787#discussion_r188655773) [](ancestors = 188655773)